services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -251,6 +257,168 @@ public URL getUrl(GetUrlRequest getUrlRequest) {
+     * bucket names that contain a dot, i.e., ".", the dot must not be URL-encoded. Encoded buckets, keys, and query parameters
+     * will be returned decoded.
+     *
+     * <p>
+     * For more information on path-style and virtual-hosted-style URIs, see <a href=
+     * "https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-bucket-intro.html"
+     * >Methods for accessing a bucket</a>.
+     *
+     * @param uri The URI to be parsed
+     * @return Parsed {@link S3Uri}
+     *
+     * <p><b>Example Usage</b>
+     * <p>
+     * {@snippet :
+     *     S3Client s3Client = S3Client.create();
+     *     S3Utilities s3Utilities = s3Client.utilities();
+     *     String uriString = "https://myBucket.s3.us-west-1.amazonaws.com/doc.txt?versionId=abc123";
+     *     URI uri = URI.create(uriString);
+     *     S3Uri s3Uri = s3Utilities.parseUri(uri);
+     *
+     *     String bucket = s3Uri.bucket().get(); // "myBucket"
+     *     String key = s3Uri.key().get(); // "doc.txt"
+     *     Region region = s3Uri.region().get(); // Region.US_WEST_1
+     *     boolean isPathStyle = s3Uri.isPathStyle(); // false
+     *     String versionId = s3Uri.firstMatchingQueryParamValue("versionId").get(); // "abc123"
+     *}
+     */
+    public S3Uri parseUri(URI uri) {
+        validateUri(uri);
+
+        if ("s3".equalsIgnoreCase(uri.getScheme())) {
+            return parseAwsCliStyleUri(uri);
+        }
+
+        return parseStandardUri(uri);
+    }
+
+    private S3Uri parseStandardUri(URI uri) {
+        String bucket = null;
+        String key = null;
+        String region = null;
+        boolean isPathStyle = false;
+        Map<String, List<String>> queryParams = new HashMap<>();
+        String path = uri.getPath();
+
+        if (uri.getHost() == null) {
+            throw new IllegalArgumentException("Invalid S3 URI: no hostname: " + uri);
+        }
+
+        Matcher matcher = ENDPOINT_PATTERN.matcher(uri.getHost());
+        if (!matcher.find()) {
+            throw new IllegalArgumentException("Invalid S3 URI: hostname does not appear to be a valid S3 endpoint: " + uri);
+        }
+
+        String prefix = matcher.group(1);
+        if (prefix == null || prefix.isEmpty()) {
+            isPathStyle = true;
+
+            if (!path.isEmpty() && !"/".equals(path)) {
+                int index = path.indexOf('/', 1);
+
+                if (index == -1) {
+                    // No trailing slash, e.g., "https://s3.amazonaws.com/bucket"
+                    bucket = path.substring(1);
+                } else {
+                    bucket = path.substring(1, index);
+                    if (index != path.length() - 1) {
+                        key = path.substring(index + 1);
+                    }
+                }
+            }
+        } else {
+            bucket = prefix.substring(0, prefix.length() - 1);
+            if (path != null && !path.isEmpty() && !"/".equals(path)) {
```

Comment:
nit, `path != null && !path.isEmpty()` can be replaced with `!StringUtils.isEmpty(path)`. Could you search other places where it can be replaced? I think I saw it somewhere else as well.

Reply:
yep, replaced here in virtual style parsing and above in path style parsing